crates/biome_dependency_graph/src/resolver_cache.rs

```diff
@@ -88,14 +92,69 @@ impl<'a> ResolverCache<'a> {
         &self.paths
     }
 
-    fn path_kind(&self, path: &Path) -> Option<PathKind> {
+    fn path_kind(&self, path: &Utf8Path) -> Option<PathKind> {
         if self.added_paths.contains(path) {
-            let utf8_path = path.try_into().ok()?;
-            self.fs.path_kind(utf8_path).ok()
+            self.fs.path_kind(path).ok()
         } else {
             self.dependency_graph.path_kind(path)
         }
     }
+
+    /// Returns the canonical path, resolving all symbolic links.
+    ///
+    /// <https://github.com/parcel-bundler/parcel/blob/4d27ec8b8bd1792f536811fef86e74a31fa0e704/crates/parcel-resolver/src/cache.rs#L232>
+    fn canonicalize_impl(&self, path: &CachedPath) -> Result<CachedPath, ResolveError> {
+        // Check if this thread is already canonicalizing. If so, we have found a circular symlink.
+        // If a different thread is canonicalizing, OnceLock will queue this thread to wait for the result.
+        let tid = THREAD_ID.with(|t| *t);
+        if path.canonicalizing.load(Ordering::Acquire) == tid {
+            return Err(io::Error::new(io::ErrorKind::NotFound, "Circular symlink").into());
+        }
+
+        path.canonicalized
+            .get_or_init(|| {
+                path.canonicalizing.store(tid, Ordering::Release);
```

Comment:
Thread ordering is quite a niche and a bit complex. Would you mind adding a comment explaining why we use `Ordering::Release`?

Reply:
This is just copied from `oxc-resolver`, which they copied from Parcel (hence the link at the top). I also didn't take the time to understand the intricacies of it. I do sometimes dabble with these `Ordering` semantics, but every time it takes me at least an hour to understand the intricacies again, so I don't think a comment is even going to help here. Either you're in a state of mind where you can follow why it works, or the comment is meaningless too :rofl: